xds/internal/xdsclient/client_refcounted.go

```diff
@@ -35,41 +34,33 @@ const (
 )
 
 var (
-	// This is the client returned by New(). It contains one client implementation,
-	// and maintains the refcount.
-	singletonMu     sync.Mutex
-	singletonClient *clientRefCounted
-
 	// The following functions are no-ops in the actual code, but can be
 	// overridden in tests to give them visibility into certain events.
-	singletonClientImplCreateHook = func() {}
-	singletonClientImplCloseHook  = func() {}
+	xdsClientImplCreateHook = func(name string) {}
+	xdsClientImplCloseHook  = func(name string) {}
 )
 
-// To override in tests.
-var bootstrapNewConfig = bootstrap.NewConfig
-
-func clientRefCountedClose() {
-	singletonMu.Lock()
-	defer singletonMu.Unlock()
+func clientRefCountedClose(name string) {
+	clientsMu.Lock()
+	defer clientsMu.Unlock()
 
-	if singletonClient.decrRef() != 0 {
+	client, ok := clients[name]
+	if !ok {
+		logger.Errorf("Attempt to close a non-existent xDS client with name %s", name)
 		return
 	}
-	singletonClient.clientImpl.close()
-	singletonClientImplCloseHook()
-	singletonClient = nil
-}
-
-func newRefCountedWithConfig(fallbackConfig *bootstrap.Config) (XDSClient, func(), error) {
-	singletonMu.Lock()
-	defer singletonMu.Unlock()
+	if client.decrRef() != 0 {
+		return
+	}
+	client.clientImpl.close()
+	xdsClientImplCloseHook(name)
+	delete(clients, name)
 
-	if singletonClient != nil {
-		singletonClient.incrRef()
-		return singletonClient, grpcsync.OnceFunc(clientRefCountedClose), nil
+}
 
-	}
+func newRefCountedWithConfig(name string, fallbackConfig *bootstrap.Config, watchExpiryTimeout, idleAuthorityTimeout time.Duration) (XDSClient, func(), error) {
```

Comment:
Top level comment of get or create if not present? Or change name?

Reply:
Same question as for the Close, is the reason to keep these names this happens implicitly with objects that are ref counted?

Reply:
Explained the reasoning in the comment for close.